specifyweb/frontend/js_src/lib/components/Resource/resourceApi.ts

```diff
@@ -0,0 +1,140 @@
+import { eventListener } from '../../utils/events';
+import { addMissingFields } from '../DataModel/addMissingFields';
+import { SerializedResource } from '../DataModel/helperTypes';
+import {
+  createResource as internalCreateResource,
+  fetchResource as internalFetchResource,
+} from '../DataModel/resource';
+import { SpecifyTable } from '../DataModel/specifyTable';
+import { tables } from '../DataModel/tables';
+import type { Tables } from '../DataModel/types';
+import { error } from '../Errors/assert';
+
+export type NewResourceRef<TABLE_NAME extends keyof Tables> = {
+  readonly table: SpecifyTable<Tables[TABLE_NAME]>;
+  readonly id: string;
+};
+
+export type SavedResourceRef<TABLE_NAME extends keyof Tables> = {
+  readonly table: SpecifyTable<Tables[TABLE_NAME]>;
+  readonly id: number;
+};
+
+export type ResourceRef<TABLE_NAME extends keyof Tables> =
+  | NewResourceRef<TABLE_NAME>
+  | SavedResourceRef<TABLE_NAME>;
+
+export type NewResource<TABLE_NAME extends keyof Tables> =
+  NewResourceRef<TABLE_NAME> & {
+    readonly ref: NewResourceRef<TABLE_NAME>;
+    readonly data: SerializedResource<Tables[TABLE_NAME]>;
+  };
+
+export type SavedResource<TABLE_NAME extends keyof Tables> =
+  SavedResourceRef<TABLE_NAME> & {
+    readonly ref: SavedResourceRef<TABLE_NAME>;
+    readonly data: SerializedResource<Tables[TABLE_NAME]>;
+  };
+
+export type Resource<TABLE_NAME extends keyof Tables> =
+  | NewResource<TABLE_NAME>
+  | SavedResource<TABLE_NAME>;
+
+const savedResourceStore: {
+  // eslint-disable-next-line prefer-readonly-type
+  [TABLE_NAME in keyof Tables]?: Record<
+    number,
+    WeakRef<SavedResource<TABLE_NAME>> | undefined
+  >;
+} = {};
+
+const newResourceStore: {
+  // eslint-disable-next-line prefer-readonly-type
+  [TABLE_NAME in keyof Tables]?:
+    | Record<string, WeakRef<NewResource<TABLE_NAME>>>
+    | undefined;
+} = {};
+
+const makeUuid = () => Math.random().toString();
+export async function fetchResource<TABLE_NAME extends keyof Tables>(
+  tableName: TABLE_NAME,
+  id: number
+): Promise<SavedResource<TABLE_NAME>> {
+  const savedResource = savedResourceStore[tableName]?.[id]?.deref();
+  if (typeof savedResource === 'object') return savedResource;
+
+  // Simplified:
+  const data = await internalFetchResource(tableName, id);
+  const ref: SavedResourceRef<TABLE_NAME> = {
+    table: tables[tableName],
+    id,
+  };
+  const resource: SavedResource<TABLE_NAME> = { ...ref, ref, data };
```

Comment:
ref includes table and id properties
it is useful to have these properties on the resource
but, it may also be useful to be able to quickly get from resource back to resource ref, so the `ref` property exists
if it turns out to be redundant/needless, we could remove it (in favor of doing `{ table: resource.table, id: resource.id }` - which could be a helper function `resourceToRef`)

note, resource refs are not globally unique - but, their contents (table name + id pair) resolve to one resource only